pallets/cash/src/chains.rs

```diff
@@ -348,72 +394,138 @@ impl FromStr for ChainId {
 #[derive(Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug, Types)]
 pub enum ChainBlock {
     Eth(<Ethereum as Chain>::Block),
+    Matic(<Polygon as Chain>::Block),
 }
 
 impl ChainBlock {
     pub fn chain_id(&self) -> ChainId {
         match self {
             ChainBlock::Eth(_) => ChainId::Eth,
+            ChainBlock::Matic(_) => ChainId::Matic,
         }
     }
 
     pub fn hash(&self) -> ChainHash {
         match self {
             ChainBlock::Eth(block) => ChainHash::Eth(block.hash),
+            ChainBlock::Matic(block) => ChainHash::Matic(block.hash),
         }
     }
 
     pub fn parent_hash(&self) -> ChainHash {
         match self {
             ChainBlock::Eth(block) => ChainHash::Eth(block.parent_hash),
+            ChainBlock::Matic(block) => ChainHash::Matic(block.parent_hash),
         }
     }
 
     pub fn number(&self) -> ChainBlockNumber {
         match self {
             ChainBlock::Eth(block) => block.number,
+            ChainBlock::Matic(block) => block.number,
         }
     }
 
+    fn map_events_eth_like<F>(
+        block: &EthereumBlock,
+        mut chain_block_event_fn: F,
+    ) -> impl Iterator<Item = ChainBlockEvent> + '_
+    where
+        F: FnMut(ChainBlockNumber, EthereumEvent) -> ChainBlockEvent + 'static,
+    {
+        block
+            .events
+            .iter()
+            .map(move |e| chain_block_event_fn(block.number, e.clone()))
+    }
+
     pub fn events(&self) -> impl Iterator<Item = ChainBlockEvent> + '_ {
-        match self {
-            ChainBlock::Eth(block) => block
-                .events
-                .iter()
-                .map(move |e| ChainBlockEvent::Eth(block.number, e.clone())),
-        }
+        let return_value: Box<dyn Iterator<Item = ChainBlockEvent>> = match self {
+            ChainBlock::Eth(block) => {
+                Box::new(Self::map_events_eth_like(block, ChainBlockEvent::Eth))
+            }
+            ChainBlock::Matic(block) => {
+                Box::new(Self::map_events_eth_like(block, ChainBlockEvent::Matic))
+            }
+        };
+
+        return_value
```

Comment:
Yeah its an extra allocate but as a `Vec<ChainBlockEvent>` should be less line noise too